src/scsi/src/main.rs

```diff
@@ -0,0 +1,392 @@
+                        status: output.status,
+                        status_qualifier: output.status_qualifier,
+                        sense: output.sense,
+                        // TODO: handle residual for data in
+                        residual: body_writer.residual(),
+                    }
+                }
+                Err(CmdError::CdbTooShort) => {
+                    // the CDB buffer is, by default, sized larger than any CDB we support; we don't
+                    // handle writes to config space (because QEMU doesn't let us), so there's no
+                    // way the guest can set it too small
+                    unreachable!();
+                }
+                Err(CmdError::DataIn(e)) => {
+                    if e.kind() == ErrorKind::WriteZero {
+                        Response::error(ResponseCode::Overrun, 0)
+                    } else {
+                        error!("Error writing response to guest memory: {}", e);
+
+                        // There's some chance the header and data in are on different descriptors,
+                        // and only the data in descriptor is bad, so let's at least try to write an
+                        // error to the header
+                        Response::error(ResponseCode::Failure, body_writer.residual())
+                    }
+                }
+            }
+        } else {
+            debug!("Rejecting command to LUN with bad target {:?}", lun);
+            Response::error(ResponseCode::BadTarget, body_writer.residual())
+        };
+
+        if let Err(e) = response.write(writer) {
+            // Alright, so something went wrong writing our response header to guest memory.
+            // The only reason this should ever happen, I think, is if the guest gave us a
+            // virtio descriptor with an invalid address.
+
+            // There's not a great way to recover from this - we just discovered that
+            // our only way of communicating with the guest doesn't work - so we either
+            // silently fail or crash. There isn't too much sense in crashing, IMO, as
+            // the guest could still recover by, say, installing a fixed kernel and
+            // rebooting. So let's just log an error and do nothing.
+            error!("Error writing response to guest memory: {:?}", e);
+        }
+    }
+
+    fn add_target(&mut self, target: Box<Target>) {
+        self.targets.push(target);
+    }
+}
+
+impl VhostUserBackendMut<VringRwLock> for VhostUserScsiBackend {
+    fn num_queues(&self) -> usize {
+        let num_request_queues = 1;
+        2 + num_request_queues
+    }
+
+    fn max_queue_size(&self) -> usize {
+        128 // qemu assumes this by default
+    }
+
+    fn features(&self) -> u64 {
+        // TODO: Any other ones worth implementing? EVENT_IDX and INDIRECT_DESC
+        // are supported by virtiofsd
+        1 << VIRTIO_F_VERSION_1 | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() | 1 << 2
```

Comment:
Yes you probably want the full set:

        1 << VIRTIO_F_VERSION_1
            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
            | 1 << VIRTIO_RING_F_INDIRECT_DESC
            | 1 << VIRTIO_RING_F_EVENT_IDX
            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()

drop the hardcoded 1 << 2